shell/common/heap_snapshot.cc

```diff
@@ -24,8 +25,10 @@ class HeapSnapshotOutputStream : public v8::OutputStream {
   void EndOfStream() override { is_complete_ = true; }
 
   v8::OutputStream::WriteResult WriteAsciiChunk(char* data, int size) override {
-    auto bytes_written = file_->WriteAtCurrentPos(data, size);
-    return bytes_written == size ? kContinue : kAbort;
+    const uint8_t* udata = reinterpret_cast<const uint8_t*>(data);
+    const size_t usize = static_cast<size_t>(std::max(0, size));
+    const auto data_span = UNSAFE_BUFFERS(base::span(udata, usize));
+    return file_->WriteAtCurrentPosAndCheck(data_span) ? kContinue : kAbort;
```

Comment:
Explanation of change:

- line 28: cast `data` to `const uint8_t` because that [`base::File::WriteAtCurrentPosAndCheck()`](https://chromium.googlesource.com/chromium/src/+/1d4147ba2f8fc394d970f67838a40915d2148387/base/files/file.h#246) needs a span of that type
- line 29: ensure `size` is non-negative because the `base::span()` constructor requires an unsigned number
- line 30: make the [base::span](https://chromium.googlesource.com/chromium/src/+/1d4147ba2f8fc394d970f67838a40915d2148387/base/containers/span.h#120). Since the [`base::span(first, count)`](https://chromium.googlesource.com/chromium/src/+/1d4147ba2f8fc394d970f67838a40915d2148387/base/containers/span.h#297) constructor is flagged `UNSAFE_BUFFER_USAGE`, our call to it must be wrapped in `UNSAFE_BUFFERS` . [context here](https://chromium.googlesource.com/chromium/src/+/59f56d97fff01d8fed1de28f4a7865d62f80ad1a). We can't avoid this here because v8's `WriteAsciiChunk()` still uses the first-and-count idiom so that's the only way we can create the span.
- line 31: replace the `base::File::WriteAtCurrentPos()` call with [`base::WriteAtCurrentPosAndCheck()`](https://chromium.googlesource.com/chromium/src/+/1d4147ba2f8fc394d970f67838a40915d2148387/base/files/file.h#246) which takes a span and checks the number of bytes written so that we don't have to.

Reply:
- Is it always safe that the data being pointed to is compatible with uint8_t?
- Could size be negative because of an error somewhere else in the code? Should KAbort if the usize is 0?
- Should we add a comment or todo for why we are using the UNSAFE_BUFFERS b/c v8::OutputStream::WriteAsciiChunk hasn't been updated to use spans for other maintainers?

Reply:
> Is it always safe that the data being pointed to is compatible with uint8_t?

Yes. Both `uint8_t` and `char` are one byte in size, so casting back & forth is safe in terms of array bounds.

The difference in use here is because V8's `WriteAsciiChunk()`'s function is basically saying "here is an ASCII string that's N bytes long" so `char` is a reasonable type. `base::File::Write*()` takes a `uint8_t` instead because it writes any data, not just strings.

> Could size be negative because of an error somewhere else in the code? Should KAbort if the usize is 0?

I don't think it is ever negative. Looking at the [calling code](https://chromium.googlesource.com/v8/v8/+/69ad9552b28299558b52eb002daf949b8f9e8bbc/src/profiler/output-stream-writer.h#113), it looks like the value passed is `v8::OutputStreamWriter::chunk_pos_` which is initialized to zero in the constructor and only ever incremented, never decremented. I think the upstream use of a signed type here is probably an oversight.

> Should we add a comment or todo for why we are using the UNSAFE_BUFFERS b/c v8::OutputStream::WriteAsciiChunk hasn't been updated to use spans for other maintainers?

Yes, this is a great idea! Added in 472592f3ef4ae0a58389c803d8e9cc173027e6fd.

Reply:
Thanks for clarifying! LGTM!